django/db/models/fields/related_descriptors.py

```diff
@@ -193,12 +193,13 @@ def __set__(self, instance, value):
         """
         # If null=True, we can assign null here, but otherwise the value needs
         # to be an instance of the related class.
+        remote_model = self.field.remote_field.model
         if value is None and self.field.null is False:
             raise ValueError(
                 'Cannot assign None: "%s.%s" does not allow null values.' %
                 (instance._meta.object_name, self.field.name)
             )
-        elif value is not None and not isinstance(value, self.field.remote_field.model):
+        elif value is not None and not isinstance(value, (remote_model, remote_model._meta.concrete_model)):
```

Comment:
Good catch, `isinstance(value, self.field.remote_field.model._meta.concrete_model)` should do in this case.